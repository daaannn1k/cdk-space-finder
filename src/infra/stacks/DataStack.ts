import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib'
import { AttributeType, ITable, Table } from 'aws-cdk-lib/aws-dynamodb'
import { Construct } from 'constructs'
import { getStackSuffix } from '../Utils'
import { Bucket, BucketAccessControl, HttpMethods, IBucket, ObjectOwnership } from 'aws-cdk-lib/aws-s3';

interface DataStackProps extends StackProps {
  distributionUrl: CfnOutput;
}

export class DataStack extends Stack {

  public readonly spacesTable: ITable;
  public readonly photosBucket: IBucket;

  constructor(scope: Construct, id: string, props?: DataStackProps) {
    super(scope, id, props)

    const stackSuffix = getStackSuffix(this);

    this.spacesTable = new Table(this, 'SpacesTable', {
      partitionKey: {
        name: 'id',
        type: AttributeType.STRING,
      },
      tableName: `SpaceStack-${stackSuffix}`,
    })

    this.photosBucket = new Bucket(this, 'SpaceFinderPhotos', {
      bucketName: `space-finder-photos-${stackSuffix}`,
      cors: [{
        allowedMethods: [HttpMethods.GET, HttpMethods.POST, HttpMethods.PUT, HttpMethods.HEAD],
        allowedOrigins: ['*'],
        allowedHeaders: ['*'],
      }],
      // accessControl: BucketAccessControl.PUBLIC_READ,
      blockPublicAccess: {
        blockPublicAcls: false,
        blockPublicPolicy: false,
        ignorePublicAcls: false,
        restrictPublicBuckets: false,
      },
      objectOwnership: ObjectOwnership.OBJECT_WRITER,
    })

    new CfnOutput(this, 'SpaceFinderPhotosBucketName', {
      value: this.photosBucket.bucketName,
    })
  }
}