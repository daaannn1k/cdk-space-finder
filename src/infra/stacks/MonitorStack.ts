import { Duration, Stack, StackProps } from 'aws-cdk-lib';
import { Alarm, Metric, Unit } from 'aws-cdk-lib/aws-cloudwatch';
import { SnsAction } from 'aws-cdk-lib/aws-cloudwatch-actions';
import { Construct } from 'constructs';
import { getStackSuffix } from '../Utils';
import { NodejsFunction } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Runtime } from 'aws-cdk-lib/aws-lambda';
import { join } from 'path';
import { Topic } from 'aws-cdk-lib/aws-sns';
import { LambdaSubscription } from 'aws-cdk-lib/aws-sns-subscriptions';




export class MonitorStack extends Stack {
  
  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)
  
    const stackSuffix = getStackSuffix(this);
    
    const webHookLambda = new NodejsFunction(this, 'WebHookLambda', {
      runtime: Runtime.NODEJS_18_X,
      handler: 'handler',
      entry: join(__dirname, '..', '..', 'services', 'monitor', 'handler.ts'),
    })

    const alarmTopic = new Topic(this, 'AlarmTopic', {
      displayName: 'AlarmTopic',
      topicName: 'AlarmTopic',
    });

    alarmTopic.addSubscription(new LambdaSubscription(webHookLambda));

    const spaceApi4xxAlarm = new Alarm(this, 'spaceApi4xxAlarm', {
      threshold: 5,
      evaluationPeriods: 1,
      metric: new Metric({
        namespace: 'AWS/ApiGateway',
        metricName: '4XXError',
        period: Duration.minutes(1),
        statistic: 'Sum',
        unit: Unit.COUNT,
        dimensionsMap: {
          'ApiName': 'SpacesApi',
        },
      }),
      alarmName: 'SpaceApi4xxAlarm',
    });

    spaceApi4xxAlarm.addAlarmAction(new SnsAction(alarmTopic));
    spaceApi4xxAlarm.addOkAction(new SnsAction(alarmTopic));
    
  }


}