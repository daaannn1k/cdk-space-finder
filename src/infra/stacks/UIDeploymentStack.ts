import { CfnOutput, Stack, StackProps } from 'aws-cdk-lib';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';
import { getStackSuffix } from '../Utils';
import { join } from 'path';
import { existsSync } from 'fs';
import { BucketDeployment, Source } from 'aws-cdk-lib/aws-s3-deployment';
import { Distribution, OriginAccessIdentity } from 'aws-cdk-lib/aws-cloudfront';
import { S3Origin } from 'aws-cdk-lib/aws-cloudfront-origins';



export class UIDeploymentStack extends Stack {

  constructor(scope: Construct, id: string, props?: StackProps) {
    super(scope, id, props)

    const stackSuffix = getStackSuffix(this)
    const deploymentBucket = new Bucket(this, 'DeploymentUIBucket', {
      bucketName: `deployment-ui-bucket-${stackSuffix}`,
    });

    const uiDir = join(__dirname, '..', '..', '..', '..', 'space-finder-fronted', 'build');

    if(!existsSync(uiDir)) {
      console.log('Such a file for this path doesnt exist' + uiDir);
      return;
    };

    new BucketDeployment(this, 'SpacesFinderDeployment', {
      destinationBucket: deploymentBucket,
      sources: [Source.asset(uiDir)]
    });

    const originIdentity = new OriginAccessIdentity(this, 'OriginAccessIdentity');

    deploymentBucket.grantRead(originIdentity);

    const distribution = new Distribution(this, 'Distribution', {
      defaultRootObject: 'index.html',
      defaultBehavior: {
        origin: new S3Origin(deploymentBucket, {
          originAccessIdentity: originIdentity
        })
      }
    });

   new CfnOutput(this, 'SpaceFinderUrl', {
      value: distribution.distributionDomainName
    });
  }
}