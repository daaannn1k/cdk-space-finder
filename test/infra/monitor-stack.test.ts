import { App } from 'aws-cdk-lib';
import { MonitorStack } from '../../src/infra/stacks/MonitorStack';
import { Capture, Match, Template } from 'aws-cdk-lib/assertions';

describe('monitor stack test', ()=> {
  let monitorStackTemplate: Template;
  beforeAll(()=> {
    const testApp = new App({
      outdir: 'cdk.out',
    });
    const monitorStack = new MonitorStack(testApp, 'testMonitorStack');
    monitorStackTemplate = Template.fromStack(monitorStack);
  })

  test('lambda properties', ()=> {
    monitorStackTemplate.hasResourceProperties('AWS::Lambda::Function', {
      Handler: 'index.handler',
      Runtime: 'nodejs18.x',
    })
  })

  test('sns topic properties', ()=> {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Topic', {
      DisplayName: 'AlarmTopic',
      TopicName: 'AlarmTopic',
    })
  })

  test('sns subscriptions with matchers', ()=> {
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription', 
    Match.objectEquals({
      Protocol: 'lambda',
      TopicArn: {
        Ref: Match.stringLikeRegexp('AlarmTopic')
      },
      Endpoint: {
        'Fn::GetAtt': [
          Match.stringLikeRegexp('WebHookLambda'),
          'Arn',
        ]
      }
    })
  )})

  test('sns subscriptions with exact values', ()=> {
    const snsTopic = monitorStackTemplate.findResources('AWS::SNS::Topic');
    const lambda = monitorStackTemplate.findResources('AWS::Lambda::Function');
    const snsTopicName = Object.keys(snsTopic)[0];
    const lambdaName = Object.keys(lambda)[0];
    
    monitorStackTemplate.hasResourceProperties('AWS::SNS::Subscription', 
      {
       Protocol: 'lambda',
        TopicArn: {
          Ref: Match.exact(snsTopicName)
        },
        Endpoint: {
          'Fn::GetAtt': [
            Match.exact(lambdaName),
            'Arn',
          ]
        }
      }
  )})

  test('alarm actions', ()=> {
    const alarmActionsCapture = new Capture()
    monitorStackTemplate.hasResourceProperties('AWS::CloudWatch::Alarm', {
      AlarmActions: alarmActionsCapture
    })

    expect(alarmActionsCapture.asArray()).toEqual([{
      Ref: expect.stringMatching(/^AlarmTopic/)
    }])
  })

  test.only('monitor stack snapshot test', ()=> {
    expect(monitorStackTemplate.toJSON()).toMatchSnapshot();
  })
})