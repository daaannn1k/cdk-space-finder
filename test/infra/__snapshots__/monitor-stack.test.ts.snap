// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`monitor stack test monitor stack snapshot test 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "AlarmTopicD01E77F9": {
      "Properties": {
        "DisplayName": "AlarmTopic",
        "TopicName": "AlarmTopic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "WebHookLambdaAlarmTopicD6696FB1": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "WebHookLambdaC2AD15FD",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "WebHookLambdaAllowInvoketestMonitorStackAlarmTopicF397B39AA86BB9BE": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "WebHookLambdaC2AD15FD",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "AlarmTopicD01E77F9",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "WebHookLambdaC2AD15FD": {
      "DependsOn": [
        "WebHookLambdaServiceRole1D0B8EE0",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "817b3d1d3e46d456785bee840de9641cfceddce82a07151fc21db3a807ff0874.zip",
        },
        "Environment": {
          "Variables": {
            "AWS_NODEJS_CONNECTION_REUSE_ENABLED": "1",
          },
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "WebHookLambdaServiceRole1D0B8EE0",
            "Arn",
          ],
        },
        "Runtime": "nodejs18.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "WebHookLambdaServiceRole1D0B8EE0": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "spaceApi4xxAlarm8E4FAA1F": {
      "Properties": {
        "AlarmActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "AlarmName": "SpaceApi4xxAlarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "SpacesApi",
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "4XXError",
        "Namespace": "AWS/ApiGateway",
        "OKActions": [
          {
            "Ref": "AlarmTopicD01E77F9",
          },
        ],
        "Period": 60,
        "Statistic": "Sum",
        "Threshold": 5,
        "Unit": "Count",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
